platform-sdk/docs/proposals/TSS-Ledger-Id/TSS-Ledger-Id.md

```diff
@@ -0,0 +1,1027 @@
+2. TSS Votes for the candidate roster must be received after the candidate roster is set.
+3. A TSS Message must exist if a TSS Vote has a 1 in the vote vector for the corresponding share index.
+
+`On Active Roster`:
+
+1. If the new active roster is the same as the existing one, do nothing.
+2. If the new active roster == the candidate roster
+   1. clear the candidate roster hash
+   2. clear the map data for the previous active roster.
+   3. remove the previous active roster hash from the voting closed set.
+3. If a previous active roster existed and the candidate roster is different from the new active roster, log an error.
+4. Set the new active roster.
+
+`On Candidate Roster`:
+
+1. If the input candidate roster is equal to the active roster, do nothing.
+2. If the input candidate roster is equal to the internal candidate roster, do nothing.
+3. Set the internal candidate roster.
+4. If a previous candidate roster existed
+   1. clear the associated `TssMessageTransaction` and `TssVoteTransaction` data.
+   2. remove the previous candidate roster hash from the voting closed set.
+
+`On TssMessageTransaction`:
+
+1. If the voting is closed, do nothing.
+2. If a `TssVoteTransaction` exists for this node, do nothing.
+3. If a TSS Message already exists for the share index and target roster, do nothing.
+4. Validate the TssMessage and if valid, add it to the list of TssMessageTransactions for the target roster.
+5. Update the sequence number bit in the vote vector with the results of validation for the TssMessageTransaction.
+6. If there are (active roster) threshold number of valid TssMessageTransactions for the target roster, send
+   the valid TssMessageTransactions transcript to the TSS Key Manager
+
+`On TssVoteTransaction`:
+
+1. If voting is closed or the vote is a duplicate of an existing vote, do nothing.
+2. Add the vote to the total and check if the threshold is met.
+3. If the threshold is met, add the roster hash to the voting closed set.
+
+`On PublicSharesAndLedgerId`:
+
+1. If the public shares and ledger id are for the candidate roster,
+   1. Sign the ledger ID with the node's private signing key.
+   2. send a `TssVoteTransaction` with the ledger id and the vote vector.
+
+##### Validate TSS Message List
+
+1. For the votes already cast in the vote vector
+   1. sum all the yes votes into a total count.
+   2. If the count is greater than or equal to the threshold in the candidate roster, do nothing and return.
+2. For each TSS message in the TSS Message list
+   1. If the TSS Message has a yes entry in the vote vector, do nothing.
+   2. otherwise, validate the message.
+   3. If the message is valid, update the vote vector and increment the count of yes votes.
+   4. If the count of yes votes is greater than or equal to the threshold in the candidate roster, then send the
+      vote vector as a system transaction and exit the validation process.
```

Comment:
```suggestion
1. For the votes already cast in the vote vector.
   1. Sum all the yes votes into a total count.
   2. If the count is greater than or equal to the threshold in the candidate roster, do nothing and return.
2. For each TSS message in the TSS Message list.
   1. If the TSS Message has a yes entry in the vote vector, do nothing.
   2. Otherwise, validate the message.
   3. If the message is valid, update the vote vector and increment the count of yes votes.
   4. If the count of yes votes is greater than or equal to the threshold in the candidate roster, then send the
      vote vector as a system transaction and exit the validation process.
```